### PR TITLE
[BEAM-8623] Add status_endpoint field to provision api ProvisionInfo

### PR DESCRIPTION
**Pull request by y1chi**
Add status_endpoint field so SDK Harness can retrieve this information.

See more: https://docs.google.com/document/d/1W77buQtdSEIPUKd9zemAM38fb-x3CvOoaTF4P2mSxmI/edit#bookmark=id.x5prmecw6cox

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by y1chi** [comment]
R: @angoenka

File: model/fn-execution/src/main/proto/beam_provision_api.proto [hunk 1]
```diff
@@ -71,6 +72,11 @@ message ProvisionInfo {
     // (required) The artifact retrieval token produced by
     // ArtifactStagingService.CommitManifestResponse.
     string retrieval_token = 6;
+
+    // (optional) The endpoint for runner to use for hosting the worker status
```
Comment (lukecwik) [inline model/fn-execution/src/main/proto/beam_provision_api.proto]: please update comment to:
```
(optional) The endpoint that the runner is hosting for the SDK to submit status reports to during pipeline execution. This field will only be populated if the runner supports SDK status reports. For more details see {link to design doc}.
```

Comment (y1chi) [reply model/fn-execution/src/main/proto/beam_provision_api.proto]: Thanks Luke for review, I've applied the change.

File: sdks/go/container/boot.go [hunk 2]
```diff
@@ -122,6 +122,10 @@ func main() {
 		"--semi_persist_dir=" + *semiPersistDir,
 		"--options=" + options,
 	}
+	if info.GetStatusEndpoint() != nil {
+		args = append(args, "--status_endpoint="+info.GetStatusEndpoint().GetUrl())
```
[the remark this answers is not in the document]
Comment (y1chi) [reply sdks/go/container/boot.go]: I think this only affects go sdk. Java and python uses env var to pass the url as I changed in their corresponding boot.go.

File: sdks/go/pkg/beam/core/runtime/harness/init/init.go [hunk 3]
```diff
@@ -42,6 +42,7 @@ var (
 	id              = flag.String("id", "", "Local identifier (required in worker mode).")
 	loggingEndpoint = flag.String("logging_endpoint", "", "Local logging gRPC endpoint (required in worker mode).")
 	controlEndpoint = flag.String("control_endpoint", "", "Local control gRPC endpoint (required in worker mode).")
+	statusEndpoint  = flag.String("status_endpoint", "", "Local status gRPC endpoint (optional in worker mode).")
```
[the remark this answers is not in the document]
Comment (y1chi) [reply sdks/go/pkg/beam/core/runtime/harness/init/init.go]: I can remove this. will there be any downsides for adding them in all together except that the variable is unused(semiPersistDir here is unused as well).

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkerHarnessHelper.java [hunk 4]
```diff
@@ -113,6 +114,18 @@ public static void configureLogging(DataflowWorkerHarnessOptions pipelineOptions
     return parseApiServiceDescriptorFromText(System.getenv().get(CONTROL_API_SERVICE_DESCRIPTOR));
   }
 
+  @Nullable
+  public static Endpoints.ApiServiceDescriptor getStatusDescriptor()
+      throws TextFormat.ParseException {
+    try {
+      return parseApiServiceDescriptorFromText(System.getenv().get(STATUS_API_SERVICE_DESCRIPTOR));
+    } catch (NullPointerException e) {
```
[the remark this answers is not in the document]
Comment (y1chi) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkerHarnessHelper.java]: good idea, thanks.

File: sdks/go/pkg/beam/core/runtime/harness/init/init.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (angoenka) [reply sdks/go/pkg/beam/core/runtime/harness/init/init.go]: Their is no downside except unused variable.

File: sdks/go/container/boot.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (angoenka) [reply sdks/go/container/boot.go]: Thanks, Noticed that we use environment variables in other places.

**Comment by angoenka** [comment]
> What about the go container?
> https://github.com/apache/beam/blob/master/sdks/go/container/boot.go

The changes are old container compatible. Go containers does not have status report as of now so instead of creating the place holder in go now, we can add it when go starts supporting status page.

**Comment by angoenka** [comment]
Please make sure that the test cases pass before committing.

**Comment by y1chi** [comment]
retest this please

**Comment by y1chi** [comment]
Run Portable_Python PreCommit

**Comment by y1chi** [comment]
Run Python PreCommit

**Comment by y1chi** [comment]
gentle ping for another look

**Comment by y1chi** [comment]
@lukecwik Luke would you mind take another look?

File: model/fn-execution/src/main/proto/beam_provision_api.proto [hunk 5]
```diff
@@ -71,6 +72,12 @@ message ProvisionInfo {
     // (required) The artifact retrieval token produced by
     // ArtifactStagingService.CommitManifestResponse.
     string retrieval_token = 6;
+
+    // (optional) The endpoint that the runner is hosting for the SDK to submit
+    // status reports to during pipeline execution. This field will only be
+    // populated if the runner supports SDK status reports. For more details see
+    // https://docs.google.com/document/d/1W77buQtdSEIPUKd9zemAM38fb-x3CvOoaTF4P2mSxmI/edit#.
```
Comment (lukecwik) [inline model/fn-execution/src/main/proto/beam_provision_api.proto]:
```suggestion
    // https://s.apache.org/beam-fn-api-harness-status
```

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkerHarnessHelper.java [hunk 6]
```diff
@@ -113,6 +115,18 @@ public static void configureLogging(DataflowWorkerHarnessOptions pipelineOptions
     return parseApiServiceDescriptorFromText(System.getenv().get(CONTROL_API_SERVICE_DESCRIPTOR));
   }
 
+  @Nullable
+  public static Endpoints.ApiServiceDescriptor getStatusDescriptor()
+      throws TextFormat.ParseException {
+    String statusApiDescriptor = System.getenv().get(STATUS_API_SERVICE_DESCRIPTOR);
+    if (Strings.isNullOrEmpty(statusApiDescriptor)) {
+      // Missing STATUS_API_SERVICE_DESCRIPTOR env var is a signal that the fn worker status api
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkerHarnessHelper.java]:
```suggestion
      // Missing STATUS_API_SERVICE_DESCRIPTOR env var is a signal that the worker status API
```

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkerHarnessHelper.java [hunk 7]
```diff
@@ -113,6 +115,18 @@ public static void configureLogging(DataflowWorkerHarnessOptions pipelineOptions
...
+      // server should be skipped.
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkerHarnessHelper.java]:
```suggestion
      // is unsupported by the current runner.
```

**Comment by y1chi** [comment]
Run Python PreCommit

**Comment by lukecwik** [comment]
Run Java PreCommit

File: sdks/go/container/boot.go (hunk 2 above)
Comment (lukecwik) [inline sdks/go/container/boot.go]:
```suggestion
		args = append(args, "--status_endpoint=" + info.GetStatusEndpoint().GetUrl())
```

File: sdks/go/pkg/beam/core/runtime/harness/init/init.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lukecwik) [reply sdks/go/pkg/beam/core/runtime/harness/init/init.go]: I think we should keep it for now

**Comment by y1chi** [comment]
Run Python PreCommit

**Comment by lukecwik** [comment]
Run Python PreCommit

**Comment by lukecwik** [comment]
Run Python PreCommit

**Comment by y1chi** [comment]
Run Python PreCommit

**Comment by y1chi** [comment]
Run Python PreCommit

**Pull request merged**